added introduction to setup and disassembling for iOS (#1464)
OWASP/owasp-mstg: Document/0x06c-Reverse-Engineering-and-Tampering.md

Because Objective-C and Swift are fundamentally different, the programming language in which the app is written affects the possibilities for reverse engineering it. For example, Objective-C allows method invocations to be changed at runtime. This makes hooking into other app functions (a technique heavily used by [Cycript](http://www.cycript.org/ "Cycript") and other reverse engineering tools) easy. This "method swizzling" is not implemented the same way in Swift, and the difference makes the technique harder to execute with Swift than with Objective-C.

> In iOS, all the application code, for both Swift and Objective-C, is compiled to machine code (e.g. ARM). Thus, to analyze iOS application a disassembler is needed. 

Reviewer: ```suggestion
On iOS, all the application code (both Swift and Objective-C) is compiled to machine code (e.g. ARM). Thus, to analyze iOS applications a disassembler is needed. 
```

> In iOS, all the application code, for both Swift and Objective-C, is compiled to native, unlike Android where code is compiled into Dalvik Bytecode and native code (if code written in C/C++). Thus, to analyze iOS application a disassembler is needed. 

Reviewer 2: A small suggestion as we say in the introduction that `native code` is c/c++. instead let's use machine code here:
```suggestion
In iOS, all the application code, for both Swift and Objective-C, is compiled to machine code (e.g. ARM). Thus, to analyze iOS application a disassembler is needed. 
```

> If you want to disassemble an application's code from the Apple Appstore, remove the Fairplay DRM first. Section "[Acquiring the App Binary](0x06b-basic-security-testing.md#acquiring-the-app-binary "Acquiring the App Binary")" in the chapter "iOS Basic Security Testing" explains how. 

Reviewer: ```suggestion
If you want to disassemble an application from the App Store, remove the Fairplay DRM first. Section "[Acquiring the App Binary](0x06b-basic-security-testing.md#acquiring-the-app-binary "Acquiring the App Binary")" in the chapter "iOS Basic Security Testing" explains how. 
```

> Before getting into how to disassemble an application's code, it is important to know that the application downloaded via Apple Appstore is encrypted using Fairplay DRM and need to be decrypted before it can be analyzed. This has been already discussed in "[Acquiring the App Binary](0x06b-basic-security-testing.md#acquiring-the-app-binary "Acquiring the App Binary")" section. In this section we assume you already have access to a decrypted iOS application.

Reviewer 2: For the overal text, try to write in the active voice and try to refer to chapter sections as part of a chapter for a better ebook/printed book :
```suggestion
If you want to disassemble an application's code from the Apple Appstore, remove the Fairplay DRM first. Section "[Acquiring the App Binary](0x06b-basic-security-testing.md#acquiring-the-app-binary "Acquiring the App Binary")" in the chapter "iOS Basic Security Testing" explains how. 
```

> In this section the term "app binary" refers to the Macho-O file in the application bundle which contains the compiled code, and should not be confused with the application bundle - the IPA file. See section "[Exploring the App Package](0x06b-basic-security-testing.md#exploring-the-app-package "Exploring the App Package")" section in chapter "Basic iOS Security Testing" for more details on the composition of IPA files. 

Reviewer: ```suggestion
In this section the term "app binary" refers to the Macho-O file in the application bundle which contains the compiled code, and should not be confused with the application bundle - the IPA file. See section "[Exploring the App Package](0x06b-basic-security-testing.md#exploring-the-app-package "Exploring the App Package")" in chapter "Basic iOS Security Testing" for more details on the composition of IPA files. 
```

> ##### Disassembling With Ghidra

Reviewer 3: Markdown lint: new line after titles :)

```suggestion
##### Disassembling With Ghidra

```

> In this section the term "app binary" refers to the Macho-O file in the application bundle which contains the compiled code, and should not be confused with the application bundle - the IPA file. Composition of an application bundle has already been discussed in "[Exploring the App Package](0x06b-basic-security-testing.md#exploring-the-app-package "Exploring the App Package")" section. 

Reviewer 2: Similar as with the previous remark:
```suggestion
In this section the term "app binary" refers to the Macho-O file in the application bundle which contains the compiled code, and should not be confused with the application bundle - the IPA file. See section "[Exploring the App Package](0x06b-basic-security-testing.md#exploring-the-app-package "Exploring the App Package")" section in chapter "Basic iOS Security Testing" for more details on the composition of IPA files. 
```

> As mentioned in "[Tooling](#tooling "Tooling")" section, Ghidra is an open source software reverse engineering tool. Among multiple functionalities offered by Ghidra, it can also be used to analyze iOS application binaries. 

Reviewer: ```suggestion
As mentioned in the "[Tooling](#tooling "Tooling")" section, Ghidra is an open source software reverse engineering tool. Among multiple functionalities offered by Ghidra, is the possibility to analyze iOS application binaries. 
```

> To get started with Ghidra is easy. Start Ghidra using `ghidraRun` (\*nix) or `ghidraRun.bat` (Windows), depending on the platform you are on. Once Ghidra is fired up, create a new project by specifying the project directory. You will be greeted by a window as shown below:  

Reviewer 3: ```suggestion
Start Ghidra using `ghidraRun` (\*nix) or `ghidraRun.bat` (Windows), depending on the platform you are on. Once Ghidra is fired up, create a new project by specifying the project directory. You will be greeted by a window as shown below:  
```

> To get started with Ghidra is easy. Start Ghidra using `ghidraRun` (\*nix) or `ghidraRun.bat` (Windows), depending on the platform you are on. Once Ghidra is fired up, create a new project by specifying the project directory. You will be greeted by a window as shown below:  

Reviewer: ```suggestion
Start Ghidra using `ghidraRun` (\*nix) or `ghidraRun.bat` (Windows), depending on the platform you are on. Once Ghidra is fired up, create a new project by specifying the project directory. You will be greeted by a window as shown below:  
```
(Removed the line because grammar was wrong, and it's subjective).

![Ghidra New Project Window](Images/Chapters/0x06c/Ghidra_new_project.png)

> In your newly created project space from above you can import the app binary by going to **File** -> **Import File** and choosing the desired file. 

Reviewer 3: ```suggestion
In your new **Active Project** you can import an app binary by going to **File** -> **Import File** and choosing the desired file. 
```

> To get started with Ghidra is easy. Start Ghidra using *ghidraRun* (\*nix) or *ghidraRun.bat* (Windows), depending on the platform you are on. Once Ghidra is fired up, create a new project by specifying the project directory. You will be greeted by a window as shown below:  

Reviewer 2: Please have a look at the [style guide](https://github.com/OWASP/owasp-mstg/blob/master/style_guide.md), maybe you can change it to:
```suggestion
To get started with Ghidra is easy. Start Ghidra using `ghidraRun` (\*nix) or `ghidraRun.bat` (Windows), depending on the platform you are on. Once Ghidra is fired up, create a new project by specifying the project directory. You will be greeted by a window as shown below:  
```

> In your newly created project space from above, you can import the app binary into this project, by going to **File** -> **Import File** and choosing the desired file. 

Reviewer: ```suggestion
In your newly created project space from above you can import the app binary by going to **File** -> **Import File** and choosing the desired file. 
```

![Ghidra import file](Images/Chapters/0x06c/Ghidra_import_binary.png)

> If the correct file is specified, Ghidra will show meta-information about the binary before starting the analysis. 

Reviewer 3: ```suggestion
If the file can be properly processed, Ghidra will show meta-information about the binary before starting the analysis. 
```

> In your newly created project space from above, you can import the app binary into this project, by going to *File* -> *Import File* and choosing the desired file. 

Reviewer 2: ```suggestion
In your newly created project space from above, you can import the app binary into this project, by going to **File** -> **Import File** and choosing the desired file. 
```

> If correct file is specified, Ghidra will show meta-information about the binary before starting the analysis. 

Reviewer 2: ```suggestion
If the correct file is specified, Ghidra will show meta-information about the binary before starting the analysis. 
```

![Ghidra Mach-O file import](Images/Chapters/0x06c/Ghidra_macho_import.png)

> To get the disassembled code for the binary file chosen above, double click the imported file. Click **yes** and **analyze** for auto-analysis on the subsequent windows. Auto-analysis will take some time depending on the size of the binary, the progress of which can be tracked in the bottom right corner of the code browser window. Once auto-analysis is completed you can start exploring the binary.  

Reviewer 3: ```suggestion
To get the disassembled code for the binary file chosen above, double click the imported file from the **Active Project** window. Click **yes** and **analyze** for auto-analysis on the subsequent windows. Auto-analysis will take some time depending on the size of the binary, the progress can be tracked in the bottom right corner of the code browser window. Once auto-analysis is completed you can start exploring the binary.  
```

> To get the disassembled code for the binary file choosen above, double click the imported file. Click **yes** and **analyze** for auto-analysis on the subsequent windows. Auto-analysis will take some time depending on the size of the binary, the progress can be tracked in the bottom right corner of the code browser window. Once auto-analysis is completed you can start exploring the binary.  

Reviewer: ```suggestion
To get the disassembled code for the binary file chosen above, double click the imported file. Click **yes** and **analyze** for auto-analysis on the subsequent windows. Auto-analysis will take some time depending on the size of the binary, the progress of which can be tracked in the bottom right corner of the code browser window. Once auto-analysis is completed you can start exploring the binary.  
```

> To get the disassembled code for the binary file choosen above, double click the imported file. Click "yes" and "analyze" for auto-analysis on the subsequent windows. Auto-analysis will take some time depending on the size of the binary, the progress can be tracked in the bottom right corner of the code browser window. Once auto-analysis is completed you can start exploring the binary.  

Reviewer 2: ```suggestion
To get the disassembled code for the binary file choosen above, double click the imported file. Click **yes** and **analyze** for auto-analysis on the subsequent windows. Auto-analysis will take some time depending on the size of the binary, the progress can be tracked in the bottom right corner of the code browser window. Once auto-analysis is completed you can start exploring the binary.  
```

![Ghidra code browser window](Images/Chapters/0x06c/Ghidra_main_window.png)

> In above annotated main window of Ghidra, disassembly window and symbol tree window are most important for moving around and exploring the binary. Decompiler window shows the decompiled version of the function selected for disassembly. It is important to note decompilation is not always accurate, but nevertheless very helpful in getting a quick understanding of the function being analyzed. It is always a good idea to keep cross-checking decompiled code against disassembled code for maximum accuracy.

Reviewer 3: I suggest to simplify the first sentences and to add the "important" note as a markdown note. I've put the GUI elements in bold according to the style guide.

```suggestion
The most important windows to explore a binary in Ghidra are the **Listing** (Disassembly) window, the **Symbol Tree** window and the **Decompiler** window, which shows the decompiled version of the function selected for disassembly.

> It is important to note that decompilation is not always accurate, but nevertheless very helpful in getting a quick understanding of the function being analyzed. It is always a good idea to keep cross-checking decompiled code against disassembled code for maximum accuracy.
```

> In above annotated main window of Ghidra, disassembly window and symbol tree window are most important for moving around and exploring the binary. Decompiler window shows the decompiled version of the function selected for disassembly. It is important to note decompilation is not always accurate, but nevertheless very helpful in getting a quick understanding of the function being analyzed. It is always a good idea to keep cross-checking decompiled code against disassembled code for maximum accuracy.

Reviewer: ```suggestion
In the main window of Ghidra, the disassembly window and symbol tree window are the most important for moving around and exploring the binary. The Decompiler window shows the decompiled version of the selected function for disassembly. It is important to note that decompilation is not always accurate, but nevertheless can be very helpful in getting a quick understanding of the analyzed function being. It is always a good idea to keep cross-checking decompiled code against disassembled code for maximum accuracy.
```

Author: Can you please recheck this suggestion. I think you mistakenly changed the original sentence "....quick understanding of the function being analyzed" to 
   "....quick understanding of the analyzed function being".

> There are many other functionalities available in Ghidra and most of them can be explored by opening the *Window* menu. For example, if you want to examine the strings present in the binary, open the **Defined Strings** option.  

Reviewer 3: style guide
```suggestion
There are many other functionalities available in Ghidra and most of them can be explored by opening the **Window** menu. For example, if you want to examine the strings present in the binary, open the **Defined Strings** option.  
```

> There are many other functionalities available in Ghidra, you can explore them by checking *Window* in menu. For example, if you want to check the strings present in the binary, which can be very helpful in reversing a binary, use **Defined Strings** option.  

Reviewer: ```suggestion
There are many other functionalities available in Ghidra and most of them can be explored by opening the *Window* menu. For example, if you want to examine the strings present in the binary, open the **Defined Strings** option.  
```

> There are many other functionalities available in Ghidra, you can explore them by checking *Window* in menu. For example, if you want to check the strings present in the binary, which can be very helpful in reversing a binary, use *Defined Strings* option.  

Reviewer 2: ```suggestion
There are many other functionalities available in Ghidra, you can explore them by checking *Window* in menu. For example, if you want to check the strings present in the binary, which can be very helpful in reversing a binary, use **Defined Strings** option.  
```

![Ghidra strings window](Images/Chapters/0x06c/Ghidra_string_window.png)

It often happens that the core logic of the application is implemented in a framework and the app binary is just a wrapper around that framework. In such cases, the above approach can be used to analyze *framework* binaries as well. 

> ##### Disassembling With IDA Pro

Reviewer 3: Markdown lint: new line after titles :)
```suggestion
##### Disassembling With IDA Pro

```

> Often the core logic of the application is implemented in the frameworks and the app binary is just a wrapper around it. In such cases, the above approach can be used to analyze *framework* binaries as well. 

Reviewer: ```suggestion
It often happens that the core logic of the application is implemented in a framework and the app binary is just a wrapper around that framework. In such cases, the above approach can be used to analyze *framework* binaries as well. 
```

> If you have license for IDA Pro, you can analyze app binary using IDA Pro as well. 

Reviewer: ```suggestion
If you have a license for IDA Pro, you can analyze the app binary using IDA Pro as well. 
```

Reviewer: Add note that this is not possible in free? (if that is the case)

> > The freeware version of IDA Pro unfortunately does not support ARM processor type.

Reviewer: ```suggestion
> The free version of IDA unfortunately does not support the ARM processor type.
```

> To get started with IDA Pro, simply open the app binary in IDA Pro. 

Reviewer 3: ```suggestion
To get started, simply open the app binary in IDA Pro. 
```

> Upon opening the file, IDA will perform auto-analysis, this is a good time to take a coffee break and check your emails. Once analysis is completed you can browse the disassembly in the disassembly window and explore functions in function window, both shown in the screenshot below. 

Reviewer: ```suggestion
Upon opening the file, IDA will perform auto-analysis, which can take a while depending on the size of the binary. Once analysis is completed you can browse the disassembly in the disassembly window and explore functions in function window, both shown in the screenshot below. 
```

> Upon opening the file, IDA will perform auto-analysis, which can take a while depending on the size of the binary. Once analysis is completed you can browse the disassembly in the disassembly window and explore functions in function window, both shown in the screenshot below. 

Reviewer 3: ```suggestion
Upon opening the file, IDA Pro will perform auto-analysis, which can take a while depending on the size of the binary. Once the auto-analysis is completed you can browse the disassembly in the **IDA View** (Disassembly) window and explore functions in the **Functions** window, both shown in the screenshot below. 
```

> IDA Pro license does not include decompiler by default and requires an additional license for Hex-Rays decompiler, which is expensive as well. On the other hand, Ghidra comes with a very capable free inbuilt decompiler, making it a compelling alternative to use for reverse engineering.  

Reviewer: ```suggestion
A normal IDA Pro license does not include a decompiler by default and requires an additional license for the Hex-Rays decompiler, which is expensive as well. On the other hand, Ghidra comes with a very capable free builtin decompiler, making it a compelling alternative to use for reverse engineering.  
```

> A normal IDA Pro license does not include a decompiler by default and requires an additional license for the Hex-Rays decompiler, which is expensive as well. On the other hand, Ghidra comes with a very capable free builtin decompiler, making it a compelling alternative to use for reverse engineering.  

Reviewer 3: ```suggestion
A regular IDA Pro license does not include a decompiler by default and requires an additional license for the Hex-Rays decompiler, which is expensive. In contrast, Ghidra comes with a very capable free builtin decompiler, making it a compelling alternative to use for reverse engineering.  
```

> If you have IDA Pro license and do not want to buy Hex-Rays decompiler, you can use Ghidra decompiler by installing [GhIDA plugin](https://github.com/Cisco-Talos/GhIDA/) for IDA. 

Reviewer: ```suggestion
If you have an IDA Pro license and do not want to buy the Hex-Rays decompiler, you can use Ghidra's decompiler by installing the [GhIDA plugin](https://github.com/Cisco-Talos/GhIDA/) for IDA. 
```

> If you have an IDA Pro license and do not want to buy the Hex-Rays decompiler, you can use Ghidra's decompiler by installing the [GhIDA plugin](https://github.com/Cisco-Talos/GhIDA/) for IDA. 

Reviewer 3: ```suggestion
If you have a regular IDA Pro license and do not want to buy the Hex-Rays decompiler, you can use Ghidra's decompiler by installing the [GhIDA plugin](https://github.com/Cisco-Talos/GhIDA/) for IDA Pro. 
```

> > The freeware version of IDA Pro unfortunately does not support ARM processor type.

Reviewer: Move this to the top.

Author: I have moved the "freeware" part to the top. Keeping the Ghida plugin at same place, as i think it is better to have after the discussion of hex-rays decompiler done in the previous paragraph.
